### PR TITLE
[ExpressionLanguage] Add support for null-safe operator

### LINKED ISSUE #45411: Expression Language package: doesn't handle the absence of the property and an error displays
### Symfony version(s) affected

^6.0

### Description

I'm using [Expression Language Package](https://github.com/symfony/expression-language) with [Laravel Framework](https://laravel.com/docs/8.x)
and I'm sending the rules object and data object to the package as mentioned in the [documentation](https://symfony.com/doc/current/components/expression_language.html#expression-syntax).

The issue is if the rules object checks property which does not exist in the data object the package displays an error.
` 
"message": "Undefined property: App\\RulesEngine\\Conditions\\DataConditions::$scores", 
"exception": "ErrorException",
 `
The idea is that I'm receiving the data object from an API and the object has different properties which not of it is required but I need to evaluate the properties in case of its existence.

I checked the package's code and I found that the code is calling the property directly without checking the existence of the property.

As example,
Rules object => 
`
data.scores = 20 || data.name = "abeer"
`
data object 
`
data['name'] = "abeer";
`
and error will appear `" Undefined property data.scores"`

### How to reproduce

**My rules engine class**
`

        class RulesEngine implements RulesEngineInterface
        {
            /** @var array<string> */
            private array $rules = [];

            public function __construct(public ExpressionLanguage $expressionLanguage)
            {
            }

            /**
            * @inheritdoc
            */
            public function addRules(array $expressions, bool $reset = false): void
            {
                $this->rules = $reset ?  [...$expressions] : [...$this->rules, ...$expressions];
            }

            /**
            * @inheritdoc
            */
            public function validateData(array $data): mixed
            {
                foreach ($this->rules as $rule) {
                    $evaluated = $this->expressionLanguage->evaluate($rule, $data);
                    if (! $evaluated) {
                        return false;
                    }
                }

                return true;
            }
        }
`
**Calling the class to handle rules evaluation**

`

        public function evaluateRules(RulesEngineInterface $rulesEngine, Game $game, array $data): bool
        {
            $rulesEngine->addRules([$game->conditions], true);

            return $rulesEngine->validateData($data);
        }
`

### Possible Solution

instead of return direct access to the property in the object [return line in Evaluate function in call GetAttrNode](https://github.com/symfony/expression-language/blob/7adc90f7f09659606996f78de22830b5a2db5be4/Node/GetAttrNode.php#L78)
check the existence of the property `object_get($obj, $property) !== null` and return false to drop only one rules without stoping the code and keep evaluating the rest of the rules.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mytuny**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45411, #21691 
| License       | MIT
| Doc PR        | symfony/symfony-docs#16630

This is a long-time-lasting feature for the `ExpressionLanguage` component. I've been waiting for the support of `Nullsafe operator` in expressions dealing with mutable objects, until I finally decided to work on it once for all 👍 

The lack of [nullsafety feature](https://wiki.php.net/rfc/nullsafe_operator) has been repeatedly reported as a BUG several time  (e.g [#45411](https://github.com/symfony/symfony/issues/45411) & [#21691](https://github.com/symfony/symfony/issues/21691)) when it is actually a missing feature.

Currently, expressions like `foo.bar` assumes that the property `bar` "always" exists on the object `foo` and if doesn't the parser throws a `RuntimeException`. Although, sometimes, that's exactly the behavior we need, some other times we may work with mutable objects with uncontrolled structure, thus, such assumption is error-prone and will force adding extra checks making the expression uglier and less readable.

The proposed work, introduces the support for the `?.` syntax alongside with the usual `.` syntax to help working with objects with dynamic structure. The two notations works identically in all normal cases. The difference occurs when trying to access non-existant properties and/or methods where the `.` notation will throw a `RuntimeException` as usual and the `?.` notation will return `null` instead and no errors nor exceptions will be thrown. Hence the name "Null-Safe".

PS: This work account ONLY for accessing **object's** properties and methods. It does not account for non-existant **array** items which is a seperate problem that can be addressed by introducing the [null coalescing](https://wiki.php.net/rfc/isset_ternary) operator. Another feature that I'm currently working on as well 💯

### CODE REVIEW

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+6.0.0
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: 6.1

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
+-----
+
+* added support of `nullsafe operator` syntax when parsing object's properties
+* added support of `nullsafe operator` syntax when parsing object's methods
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: Add instead of added. Let's merge both lines also

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php [hunk 3]
```diff
@@ -34,19 +34,20 @@ public function __construct(Node $node, Node $attribute, ArrayNode $arguments, i
 
     public function compile(Compiler $compiler)
     {
+        $nullSafe = $this->nodes['attribute'] instanceof ConstantNode ? $this->nodes['attribute']->isNullSafe : false;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: `this->nodes['attribute'] instanceof ConstantNode && $this->nodes['attribute']->isNullSafe;`

**Comment by mytuny** [comment]
I start working on this PR before the 6.1 release, therefore, this work is based on branch 5.4 instead.

File: src/Symfony/Component/ExpressionLanguage/Parser.php [hunk 4]
```diff
@@ -331,11 +331,19 @@ public function parseHashExpression()
         return $node;
     }
 
-    public function parsePostfixExpression(Node\Node $node)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: Because the method is public, adding an argument is a bc break. We should use a virtual argument here, and read it using func_get_arg. See similar logic in the codebase, mostly on branch *.4

**Review by nicolas-grekas — COMMENTED** [review]
That's great thanks! Here are some minor comments

**Comment by nicolas-grekas** [comment]
Right, please rebase to target ~5.4~6.1 as that's the branch where new features are merged.

**Comment by fabpot** [comment]
@nicolas-grekas wanted to say 6.1, not 5.4, for the branch :)

**Comment by mytuny** [comment]
I just rebased to 6.1, however, this seem to pollute the PR commits history :/ + had to resolve few conflits.
Don't you think that is better to re-open a clean PR based on 6.1 branch ?

**Comment by stof** [comment]
@mytuny this polluted it, because you were still targetting the 5.4 branch in the PR. I edited the PR to change the target branch to 6.1, which fixed it.

**Comment by stof** [comment]
also, it looks like you merged your rebased branch with the upstream branch instead of forcing the push, thus duplicating the commits in the history. See https://symfony.com/doc/current/contributing/code/pull_requests.html#rebase-your-pull-request

**Comment by mytuny** [comment]
> @mytuny this polluted it, because you were still targetting the 5.4 branch in the PR. I edited the PR to change the target branch to 6.1, which fixed it.

Brilliant! Many thanks 👏

**Comment by mytuny** [comment]
> also, it looks like you merged your rebased branch with the upstream branch instead of forcing the push, thus duplicating the commits in the history. See https://symfony.com/doc/current/contributing/code/pull_requests.html#rebase-your-pull-request

Although I followed exactly the steps mentioned there, a conflit-resolving commits that I've added during the rebase of 6.1 has introduced that duplication. The conflits occured with files that have been changed both in 6.1 and in my PR as well comparing to 5.4, namely the CHANGELOG file and the ConstantNode.php file. That's why I asked if it is better to re-submit a PR based on 6.1 right from the start 🤔

**Comment by stof** [comment]
@mytuny the process I linked to is using a force push, precisely to **avoid** doing a merge of the remote branch after the rebase.
You should not have the commit `Merge remote-tracking branch 'origin/feature_expressionlanguage_nulls` at all in your branch.

**Comment by mytuny** [comment]
> 

Yess, exactly :( any suggestions ?

**Comment by stof** [comment]
@mytuny use `git reset --hard bffdcd22cabac5e105697b44e8e34c080dbe5f39` to reset your branch to the commit before the bad merge, and then actually follow the end of the process as documented

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md [hunk 5]
```diff
@@ -1,11 +1,13 @@
 CHANGELOG
 =========
 
-6.1
+6.1.0
```
Comment (OskarStark) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md:4]:
```suggestion
6.1
```

**Comment by mytuny** [comment]
@stof All right! the hard reset does the trick.
This time I forced the push and we seem to have a cleaner commits history by now...
Thank you :)

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: should be reverted :)

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
 -----
 
- * Support lexing numbers with the numeric literal separator `_`
- * Support lexing decimals with no leading zero
+* add support for `nullsafe operator` syntax when parsing object's properties
+* add support for `nullsafe operator` syntax when parsing object's methods
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: Add instead of add + let's merge both lines

File: src/Symfony/Component/ExpressionLanguage/Node/ConstantNode.php [hunk 7]
```diff
@@ -20,11 +20,14 @@
  */
 class ConstantNode extends Node
 {
+
     private bool $isIdentifier;
+    public bool $isNullSafe;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/ConstantNode.php]: public should be first, and blank line above should be removed
let's declare it as readonly also

File: src/Symfony/Component/ExpressionLanguage/Parser.php [hunk 8]
```diff
@@ -356,7 +364,8 @@ public function parsePostfixExpression(Node\Node $node)
                     throw new SyntaxError('Expected name.', $token->cursor, $this->stream->getExpression());
                 }
 
-                $arg = new Node\ConstantNode($token->value, true);
+                $isNullSafe = func_num_args() >= 2 && func_get_arg(1);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: actually, looking at the code more carefully, this new argument should not be needed. We can remove it if we don't recurse into parsePostfixExpression on L339. That'd be better.

File: src/Symfony/Component/ExpressionLanguage/Parser.php (hunk 4 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/Parser.php]: [Done!](https://github.com/symfony/symfony/pull/45795/commits/bffdcd22cabac5e105697b44e8e34c080dbe5f39)

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md (hunk 1 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: [Done!](https://github.com/symfony/symfony/pull/45795/commits/bffdcd22cabac5e105697b44e8e34c080dbe5f39)

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md (hunk 2 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: [Done!](https://github.com/symfony/symfony/pull/45795/commits/bffdcd22cabac5e105697b44e8e34c080dbe5f39)

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php (hunk 3 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: [Done!](https://github.com/symfony/symfony/pull/45795/commits/bffdcd22cabac5e105697b44e8e34c080dbe5f39)

File: src/Symfony/Component/ExpressionLanguage/Parser.php (hunk 8 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/Parser.php]: I agree! I smell a place of improvment over there .. let me think about it 🤔

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: `Add support for null-safe syntax when parsing object's methods and properties`

File: src/Symfony/Component/ExpressionLanguage/CHANGELOG.md [hunk 9]
```diff
@@ -1,11 +1,13 @@
...
+* Support lexing numbers with the numeric literal separator `_`
+* Support lexing decimals with no leading zero
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/CHANGELOG.md]: oh btw, a space is needed at start of line

File: src/Symfony/Component/ExpressionLanguage/Parser.php (hunk 8 above)
Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/Parser.php]: I just suggested a small refactoing of the `parsePostfixExpression` method which is BC-safe and at the same time avoids recursing on itself for the sake of applying the null-safety condition. The trick is to isolate the logic of object-access with the `.` punctuation (see the method `parseObjectAccessExpression`) since it will be the common re-usable logic in both scenarios. Same solution can be applied to the array-access logic with the `[` ponctuation (see the method `parseArrayAccessExpression`) to prepare the field for any future "special" mechanism to be seperately added there, for instance, we can add the null-safe support for array-items-access as well (like `?[` 🙄)  if we choose too!! (a feature even PHP8 itself does not support ^^).

File: src/Symfony/Component/ExpressionLanguage/Node/ConstantNode.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/ExpressionLanguage/Node/ConstantNode.php]: `public readonly bool $isNullSafe;`

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php [hunk 10]
```diff
@@ -75,6 +76,10 @@ public function evaluate(array $functions, array $values)
 
                 $property = $this->nodes['attribute']->attributes['value'];
 
+                if ($this->nodes['attribute']->isNullSafe) {
+                    return $obj->$property ?? null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: wait, this is not what nullsafe means. nullsafe is about the left side being null (the object), not the right side (the property)
same issue on the method call below

I guess this needs to be covered by a test case

Comment (mytuny) [reply src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: We can simply do this:
```suggestion
                    return $obj?->$property;
```
But then we should exclude throwing a `RuntimeException` from line 73 above it when we use the null-safe syntax!
like this:
```
                if (!\is_object($obj) && !$this->nodes['attribute']->isNullSafe) {
                    throw new \RuntimeException(sprintf('Unable to get property "%s" of non-object "%s".', $this->nodes['attribute']->dump(), $this->nodes['node']->dump()));
                }
```

Comment (nicolas-grekas) [reply src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: $obj cannot be null here so that won't help

File: src/Symfony/Component/ExpressionLanguage/Tests/ExpressionLanguageTest.php [hunk 11]
```diff
@@ -245,6 +245,13 @@ public function testCallBadCallable()
         $el->evaluate('foo.myfunction()', ['foo' => new \stdClass()]);
     }
 
+    public function testCallBadCallableWithNullSafe()
+    {
+        $el = new ExpressionLanguage();
+        $result = $el->evaluate('foo?.myfunction()', ['foo' => new \stdClass()]);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Tests/ExpressionLanguageTest.php]: This test is not correct: `foo?.myfunction()` should mean that if `foo` is null or undefined, then `null` should be returned.
this is unrelated to the method existing or not

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php [hunk 12]
```diff
@@ -70,16 +71,26 @@ public function evaluate(array $functions, array $values)
...
                 if (!\is_object($obj)) {
+                    if (is_null($obj) && $this->nodes['attribute']->isNullSafe) {
+                        return null;
+                    }
                     throw new \RuntimeException(sprintf('Unable to get property "%s" of non-object "%s".', $this->nodes['attribute']->dump(), $this->nodes['node']->dump()));
                 }
 
                 $property = $this->nodes['attribute']->attributes['value'];
 
+                if (!property_exists($obj, $property)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: we should remove this since now it's unrelated (and might break dynamic properties implemented with __get())

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php [hunk 13]
```diff
@@ -70,16 +71,26 @@ public function evaluate(array $functions, array $values)
             case self::PROPERTY_CALL:
                 $obj = $this->nodes['node']->evaluate($functions, $values);
                 if (!\is_object($obj)) {
+                    if (is_null($obj) && $this->nodes['attribute']->isNullSafe) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: `if (null === $obj [...]` + I'd suggest to move this "if" outside of the parent "if"

File: src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php [hunk 14]
```diff
@@ -70,16 +71,26 @@ public function evaluate(array $functions, array $values)
...
+                if (!property_exists($obj, $property)) {
+                    throw new \RuntimeException(sprintf('Unable to get property "%s" of object "%s".', $this->nodes['attribute']->dump(), $this->nodes['node']->dump()));
+                }
+
                 return $obj->$property;
 
             case self::METHOD_CALL:
                 $obj = $this->nodes['node']->evaluate($functions, $values);
                 if (!\is_object($obj)) {
+                    if (is_null($obj) && $this->nodes['attribute']->isNullSafe) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/GetAttrNode.php]: same as above

File: src/Symfony/Component/ExpressionLanguage/Tests/ExpressionLanguageTest.php [hunk 15]
```diff
@@ -237,6 +237,37 @@ public function testRegisterAfterEval($registerCallback)
...
+        $this->expectException(\RuntimeException::class);
+        $this->expectExceptionMessageMatches('/Unable to get property "\w+" of non-object "\w+"./');
+        $el = new ExpressionLanguage();
+        $el->evaluate('foo.bar', ['foo' => null]);
+    }
+
+    public function testBadObjectGetPropertyNullSafe()
+    {
+        $el = new ExpressionLanguage();
+        $result = $el->evaluate('foo?.bar', ['foo' => null]);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Tests/ExpressionLanguageTest.php]: please add a test case were the foo variable is not declared,
and also one where the nullsafe is nested foo.bar?.baz, where only foo exists and is not null

**Review by nicolas-grekas — COMMENTED** [review]
almost good to me :)

**Review by nicolas-grekas — APPROVED** [review]
I just fixed my own comments and rebased. LGTM :+1:

**Comment by nicolas-grekas** [comment]
Note that the nullsafe operator only guards against `null`. It does not guard against undefined properties/keys.

**Comment by nicolas-grekas** [comment]
Thank you @mytuny.

**Comment by nicolas-grekas** [comment]
> Note that the nullsafe operator only guards against null. It does not guard against undefined properties/keys.

About this, we might want to add support for the null-coalescing operator `??`, which is able to ignore undefined vars/properties/keys.

@mytuny up to give this a try as a follow up? Please have a look at the changes I made they might help (esp. the part where the lexer is modified);

**Comment by mytuny** [comment]
Thank you @nicolas-grekas !! 
I'm so glad to help improve this incredible project :)

**Comment by mytuny** [comment]
Thank you @nicolas-grekas !! 
I'm so glad to help improve this incredible project :)

**Comment by mytuny** [comment]
> > Note that the nullsafe operator only guards against null. It does not guard against undefined properties/keys.
> 
> About this, we might want to add support for the null-coalescing operator `??`, which is able to ignore undefined vars/properties/keys.
> 
> @mytuny up to give this a try as a follow up? Please have a look at the changes I made they might help (esp. the part where the lexer is modified);

Yup! sure, I'm working on that null-coalescing operator right now. Your comments did and will help me even more to have the feature ready for the next release too.

Thank you once more @nicolas-grekas
